File: webui_utilities/pages/base_page.py
import allure
from playwright.sync_api import Page
from typing import Optional


class BasePage:
    """Base class for all page objects."""

    def __init__(self, page: Page):
        self.page = page

    def wait_for_element(self, selector: str, timeout: int = 3000):
        """Wait for an element to be visible."""
        with allure.step(f"Wait for element with selector: {selector}"):
            self.page.wait_for_selector(selector, timeout=timeout)

    def click(self, selector: str):
        """Click on an element."""
        with allure.step(f"Click element: {selector}"):
            self.page.click(selector)

    def fill(self, selector: str, text: str):
        """Fill text input."""
        with allure.step(f"Fill field: {selector} with text: {text}"):
            self.page.fill(selector, text)

    def get_text(self, selector: str) -> str:
        """Get text from an element."""
        with allure.step(f"Get text from element: {selector}"):
            return self.page.text_content(selector)

    def is_element_visible(self, selector: str) -> bool:
        """Check if element is visible."""
        try:
            self.page.wait_for_selector(selector, timeout=1000)
            return True
        except:
            return False

    def get_attribute(self, selector: str, attr: str) -> Optional[str]:
        """Get element attribute."""
        with allure.step(f"Get attribute '{attr}' from element: {selector}"):
            return self.page.get_attribute(selector, attr)

    def select_option(self, selector: str, value: str):
        """Select option from dropdown."""
        with allure.step(f"Select option '{value}' from dropdown: {selector}"):
            self.page.select_option(selector, value)

    def take_screenshot(self, name: str):
        """Take a screenshot."""
        with allure.step(f"Take screenshot: {name}"):
            self.page.screenshot(path=f"screenshots/{name}.png")

File: src/pages/BasePage.ts
import { Page, expect } from '@playwright/test';
import { config } from '../utils/config';

/**
 * Base Page Object Model class for all pages
 * Provides common methods for interacting with web elements
 */
export class BasePage {
  constructor(protected page: Page) {}

  /**
   * Navigate to a URL
   */
  async navigate(url: string): Promise<void> {
    await this.page.goto(url, { waitUntil: 'domcontentloaded' });
  }

  /**
   * Wait for element to be visible
   */
  async waitForElement(selector: string, timeout: number = config.elementWaitTimeout): Promise<void> {
    await this.page.waitForSelector(selector, { timeout });
  }

  /**
   * Click on element
   */
  async click(selector: string): Promise<void> {
    await this.page.locator(selector).click();
  }

  /**
   * Fill text input or select element
   */
  async fill(selector: string, text: string): Promise<void> {
    const element = this.page.locator(selector);
    const tagName = await element.evaluate((el) => el.tagName.toLowerCase());
    
    if (tagName === 'select') {
      await element.selectOption(text);
    } else {
      await element.fill(text);
    }
  }

  /**
   * Get text content of element
   */
  async getText(selector: string): Promise<string | null> {
    return await this.page.textContent(selector);
  }

  /**
   * Get attribute value of element
   */
  async getAttribute(selector: string, attribute: string): Promise<string | null> {
    return await this.page.getAttribute(selector, attribute);
  }

  /**
   * Check if element is visible
   */
  async isElementVisible(selector: string): Promise<boolean> {
    try {
      await this.page.waitForSelector(selector, { timeout: config.elementWaitTimeout });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Check if element is hidden
   */
  async isElementHidden(selector: string): Promise<boolean> {
    try {
      await this.page.waitForSelector(selector, { state: 'hidden', timeout: config.elementWaitTimeout });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get all text content from elements matching selector
   */
  async getAllText(selector: string): Promise<string[]> {
    return await this.page.locator(selector).allTextContents();
  }

  /**
   * Select option from dropdown
   */
  async selectOption(selector: string, value: string): Promise<void> {
    await this.page.selectOption(selector, value);
  }

  /**
   * Check element is enabled
   */
  async isElementEnabled(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isEnabled();
  }

  /**
   * Check element is disabled
   */
  async isElementDisabled(selector: string): Promise<boolean> {
    return !(await this.isElementEnabled(selector));
  }

  /**
   * Hover over element
   */
  async hover(selector: string): Promise<void> {
    await this.page.hover(selector);
  }

  /**
   * Get page title
   */
  async getPageTitle(): Promise<string> {
    return await this.page.title();
  }

  /**
   * Get current URL
   */
  async getCurrentUrl(): Promise<string> {
    return this.page.url();
  }

  /**
   * Wait for navigation
   */
  async waitForNavigation(): Promise<void> {
    await this.page.waitForNavigation({ timeout: config.navigationTimeout });
  }

  /**
   * Reload page
   */
  async reloadPage(): Promise<void> {
    await this.page.reload();
  }
}
